tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,342 @@
+	var (
+		err       error
+		leaseName string
+	)
+
+	if containerName != nginxContainerName {
+		// Since we have already deployed resources and ran resourceManager.WaitForAppsToBeReady(ns.Name) earlier,
+		// we know that the applications are ready at this point. This could only be the case if NGF has written
+		// statuses, which could only be the case if NGF has the leader lease. Since there is only one instance
+		// of NGF in this test, we can be certain that this is the correct leaseholder name.
+		leaseName, err = getLeaderElectionLeaseHolderName()
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	restartContainer(ngfPodName, containerName)
+
+	checkContainerLogsForErrors(ngfPodName)
+
+	if containerName != nginxContainerName {
+		Eventually(
+			func() bool {
+				return checkLeaderLeaseChange(leaseName)
+			}).
+			WithTimeout(timeoutConfig.GetLeaderLeaseTimeout).
+			WithPolling(500 * time.Millisecond).
+			Should(BeTrue())
+	}
+
+	Eventually(
+		func() bool {
+			return checkForWorkingTraffic(teaURL, coffeeURL)
+		}).
+		WithTimeout(timeoutConfig.RequestTimeout).
+		WithPolling(500 * time.Millisecond).
+		Should(BeTrue())
+
+	Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+
+	Eventually(
+		func() bool {
+			return checkForFailingTraffic(teaURL, coffeeURL)
+		}).
+		WithTimeout(timeoutConfig.RequestTimeout).
+		WithPolling(500 * time.Millisecond).
+		Should(BeTrue())
+
+	Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+	Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+	Eventually(
+		func() bool {
+			return checkForWorkingTraffic(teaURL, coffeeURL)
+		}).
+		WithTimeout(timeoutConfig.RequestTimeout).
+		WithPolling(500 * time.Millisecond).
+		Should(BeTrue())
+}
+
+func restartContainer(ngfPodName, containerName string) {
+	var jobScript string
+	if containerName == "nginx" {
+		jobScript = "PID=$(pgrep -f \"nginx: master process\") && kill -9 $PID"
+	} else {
+		jobScript = "PID=$(pgrep -f \"/usr/bin/gateway\") && kill -9 $PID"
+	}
+
+	restartCount, err := getContainerRestartCount(ngfPodName, containerName)
+	Expect(err).ToNot(HaveOccurred())
+
+	job, err := runNodeDebuggerJob(ngfPodName, jobScript)
+	Expect(err).ToNot(HaveOccurred())
+
+	Eventually(
+		func() bool {
+			return checkContainerRestart(ngfPodName, containerName, restartCount)
+		}).
+		WithTimeout(timeoutConfig.ContainerRestartTimeout).
+		WithPolling(500 * time.Millisecond).
+		Should(BeTrue())
+
+	// default propagation policy is metav1.DeletePropagationOrphan which does not delete the underlying
+	// pod created through the job after the job is deleted. Setting it to metav1.DeletePropagationBackground
+	// deletes the underlying pod after the job is deleted.
+	Expect(resourceManager.Delete(
+		[]client.Object{job},
+		client.PropagationPolicy(metav1.DeletePropagationBackground),
+	)).ToNot(HaveOccurred())
+}
+
+func checkContainerRestart(ngfPodName, containerName string, currentRestartCount int) bool {
```

Comment:
In checks like this, it's better to return an error with a description of a problem. This way if Eventually(...).To(Succeed()) fails, it will print that error, rather than printing `expected true, got false`. Otherwise, when a test doesn't work, it will not be clear why.

For example, in `checkForWorkingTraffic`, there might be some specific communication problem with NGF, but we will not be able to see it in the test output.

Note: I realize my original suggestion involved `bool`, but I just realized that it is better to have error when writing something similar in my test today.